src/util/mmap_impl_win32.h

```diff
@@ -0,0 +1,246 @@
+                                    << filename_ << " to " << file_end
+                                    << "bytes";
+        }
+      }
+      map_access = FILE_MAP_WRITE; //read/write access
+      protection = PAGE_READWRITE;
+      break;
+  }
+
+  const auto mapping = CreateFileMappingW(this->fd_, nullptr, protection,
+                                          0, 0, nullptr);
+  if (mapping == NULL) {
+    return JPPS_INVALID_STATE << "Cannot open file mapping for "
+                              << filename_
+                              << ". Error: " << GetLastError();
+  }
+
+  const DWORD granuality = get_system_granuality();
+
+  const size_t alignment = offset % static_cast<size_t>(get_system_granuality());
+  const size_t aligned_offset = offset - alignment;
+  const size_t aligned_len = size + alignment;
+
+  const auto map_view = MapViewOfFile(mapping, map_access,
+                                      static_cast<DWORD>(aligned_offset >> 16 >> 16),
+                                      static_cast<DWORD>(aligned_offset & 0xffffffff),
+                                      aligned_len);
+  CloseHandle(mapping);
+
+  if (map_view == NULL) {
+    return JPPS_INVALID_STATE << "Cannot open mapping view for "
+                              << filename_
+                              << ". Error: " << GetLastError();
+  }
+
+  view->address_ = static_cast<char*>(map_view) + alignment;
+  view->size_ = size;
+
+  return Status::Ok();
+}
+
+void MappedFile::close() {
+  if (fd_ != nullptr) {
+    CloseHandle(fd_);
+    fd_ = nullptr;
+    filename_.clear();
+    size_ = 0;
+  }
+}
+
+MappedFileFragment::MappedFileFragment() noexcept : address_{nullptr} {}
+
+MappedFileFragment::~MappedFileFragment() {
+  if (!isClean()) {
+    this->unmap();
+  }
+}
+
+void MappedFileFragment::unmap() {
+  const size_t alignment = reinterpret_cast<size_t>(this->address_)
+                           % static_cast<size_t>(get_system_granuality());
+  const void *ptr = static_cast<char*>(this->address_) - alignment;
+  UnmapViewOfFile(ptr);
+  address_ = nullptr;
+}
+
+bool MappedFileFragment::isClean() { return address_ == nullptr; }
+
+Status MappedFileFragment::flush() {
+  if (FlushViewOfFile(this->address_, 0)) {
+    return Status::InvalidState()
+           << "could not flush mapped contents, error: " << GetLastError();
+  }
+  return Status::Ok();
+}
+StringPiece MappedFileFragment::asStringPiece() const {
+  StringPiece::pointer_t asChar =
+      reinterpret_cast<StringPiece::pointer_t>(address_);
+  return StringPiece(asChar, size_);
+}
+
+MappedFileFragment::MappedFileFragment(MappedFileFragment &&o) noexcept
+    : address_{o.address_}, size_{o.size_} {
+  o.address_ = nullptr;
+}
+
+MappedFileFragment &MappedFileFragment::operator=(
+    MappedFileFragment &&o) noexcept {
+  address_ = o.address_;
+  size_ = o.size_;
+  o.address_ = nullptr;
+  return *this;
+}
+
+Status FullyMappedFile::open(StringPiece filename, MMapType type) {
```

Comment:
This logic seems to be completely duplicated between win/unix impls

Reply:
Definitely, will move it to `mmap.cc`